File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/AuthenticateUserUseCaseImpl.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.application.dtos.LoginCommand;
import com.example.orderservice.application.dtos.LoginResult;
import com.example.orderservice.domain.events.UserLoggedIn;
import com.example.orderservice.domain.models.*;
import com.example.orderservice.domain.ports.EventPublisher;
import com.example.orderservice.domain.ports.PasswordHasher;
import com.example.orderservice.domain.ports.TokenGenerator;
import com.example.orderservice.domain.ports.UserRepository;

public class AuthenticateUserUseCaseImpl implements AuthenticateUserUseCase {

    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;
    private final TokenGenerator tokenGenerator;
    private final EventPublisher eventPublisher;

    public AuthenticateUserUseCaseImpl(
            UserRepository userRepository,
            PasswordHasher passwordHasher,
            TokenGenerator tokenGenerator,
            EventPublisher eventPublisher) {
        this.userRepository = userRepository;
        this.passwordHasher = passwordHasher;
        this.tokenGenerator = tokenGenerator;
        this.eventPublisher = eventPublisher;
    }

    @Override
    public LoginResult execute(LoginCommand command) {
        if (command == null) {
            throw new IllegalArgumentException("command must not be null");
        }

        Email email = new Email(command.email());
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new AuthenticationException("AUTH_INVALID_CREDENTIALS",
                "Invalid email or password"));

        if (!user.authenticate(command.password(), passwordHasher)) {
            throw new AuthenticationException("AUTH_INVALID_CREDENTIALS",
                "Invalid email or password");
        }

        user.recordLogin();
        userRepository.save(user);

        String accessToken = tokenGenerator.generateAccessToken(user);
        String refreshToken = tokenGenerator.generateRefreshToken(user);

        eventPublisher.publish(new UserLoggedIn(user.getId()));

        return new LoginResult(
            accessToken,
            refreshToken,
            user.getEmail().getValue(),
            user.getRoles(),
            "Bearer"
        );
    }
}

File: boilerplate/nestjs/order-service/src/application/usecases/authenticate-user.use-case.impl.ts
import { Injectable, Inject } from '@nestjs/common';
import { IAuthenticateUserUseCase } from './authenticate-user.use-case.interface';
import { LoginCommand, LoginResult } from '../dtos/auth.dto';
import { IUserRepository } from '../../domain/ports/user-repository.port';
import { IPasswordHasher } from '../../domain/ports/password-hasher.port';
import { ITokenGenerator } from '../../domain/ports/token-generator.port';
import { IEventPublisher } from '../../domain/ports/event-publisher.port';
import { Email } from '../../domain/models/email.value-object';
import { AuthenticationException } from '../../domain/exceptions/auth.exception';
import { UserLoggedInEvent } from '../../domain/events/user-logged-in.event';

@Injectable()
export class AuthenticateUserUseCaseImpl implements IAuthenticateUserUseCase {
    constructor(
        @Inject('IUserRepository')
        private readonly userRepository: IUserRepository,
        @Inject('IPasswordHasher')
        private readonly passwordHasher: IPasswordHasher,
        @Inject('ITokenGenerator')
        private readonly tokenGenerator: ITokenGenerator,
        @Inject('IEventPublisher')
        private readonly eventPublisher: IEventPublisher,
    ) {}

    async execute(command: LoginCommand): Promise<LoginResult> {
        if (!command) {
            throw new Error('Command must not be null');
        }

        const email = new Email(command.email);
        const user = await this.userRepository.findByEmail(email);

        if (!user) {
            throw new AuthenticationException('AUTH_INVALID_CREDENTIALS', 'Invalid email or password');
        }

        const isAuthenticated = await user.authenticate(command.password, this.passwordHasher);

        if (!isAuthenticated) {
            throw new AuthenticationException('AUTH_INVALID_CREDENTIALS', 'Invalid email or password');
        }

        user.recordLogin();
        await this.userRepository.save(user);

        const accessToken = this.tokenGenerator.generateAccessToken(user);
        const refreshToken = this.tokenGenerator.generateRefreshToken(user);

        await this.eventPublisher.publish(new UserLoggedInEvent(user.id.getValue()));

        return {
            accessToken,
            refreshToken,
            email: user.email.getValue(),
            roles: user.roles,
            tokenType: 'Bearer',
        };
    }
}
